### ISSUE TITLE
NPCs may spawn with some of their items next to them on the ground

### ISSUE THREAD

**Issue by NetSysFire** [CleverRaven/Cataclysm-DDA#65507]
### Describe the bug

![image](https://user-images.githubusercontent.com/59517351/236623149-1d94d521-4fe2-4f5f-8929-96f19dc82405.png)

![image](https://user-images.githubusercontent.com/59517351/236623228-e5e278b7-6827-4646-8bb6-3bd161077934.png)

### Attach save file

n/a

### Steps to reproduce

1. Have an empty area to test in, e.g a field.
2. Spawn NPCs via the debug menu > Spawning > Spawn NPC
3. Repeat the step until you see a pile of items on the ground right next to them. It might take a lot of tries. You might also want to move a few tiles after spawning each of them for ease of detection of piles.

### Expected behavior

The NPCs being able to hold all their items. I was told on Discord that it might be a case of improperly set up item groups.

### Versions and configuration

- OS: Linux
    - OS Version: LSB Version: n/a; Distributor ID: Arch; Description: Arch Linux; Release: rolling; Codename: n/a; 
- Game Version: 40a2e5f [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]

**Comment by RenechCDDA** [comment]
I saw the conversation in the development discord, but this looks to be *partly* item group related. Check out this reproduction:

![image](https://user-images.githubusercontent.com/84619419/236623547-526c5cb3-18ee-440f-a03f-43296fb05c6b.png)

Of course the wooden barrel(110L) won't fit in any reasonable pocket, but the glass jar? That's suspicious. Going on that I repeated for a second reproduction....

![image](https://user-images.githubusercontent.com/84619419/236623593-8b8ba6e8-fd15-41a5-bcfb-eafd3e3ea7df.png)

A 9mm pistol and its related ammo! Aha. I have a feeling it's dropping everything of a given nested group, or something along those lines. (Or perhaps it's an ordering problem, i.e. if there's 17 items to put in and # 15 is oversized it'll drop 15, 16, 17)

**Comment by RenechCDDA** [comment]
Okay so this all ends up being about `npc::randomize`.

https://github.com/CleverRaven/Cataclysm-DDA/blob/d3ea970c8d37cdfe52942e8a16a6b2da76f27a42/src/npc.cpp#L847-L849

So we set the starting item, then their clothes, then... their inventory.

 `npc::starting_weapon` expects for a NPC with highest skill of pistol to be getting one weapon from `Item_spawn_data_guns_pistol_common` (so the guns_pistol_common group), but that in fact ends up being an entire gaggle of items. In one case I got a taurus spectrum, which can only have come from nested_taurus_spectrum which itself is only in guns_pistol_common. That's set because the NPC's highest skill is pistols at 7. 

https://github.com/CleverRaven/Cataclysm-DDA/blob/d3ea970c8d37cdfe52942e8a16a6b2da76f27a42/src/npc.cpp#L1264-L1265

Hm, okay. But why is it on the **ground**?

starting_inv immediately clears the inventory - but the NPC is still technically *wielding* that entire item group. Uh, okay, so unaffected. starting_inv continues to do things like... put ammo into lighters and otherwise make sure the wielded weapon's ammo matches, etc. And then we get down to the problem.

https://github.com/CleverRaven/Cataclysm-DDA/blob/d3ea970c8d37cdfe52942e8a16a6b2da76f27a42/src/npc.cpp#L1080-L1094

Basically, it randomizes a number (never less than 2, but up to 15 at max) and then gives that amount of random items from "<class-name>_<what>" or from "npc_<what>"... (so the `npc_misc` item group but that isn't the cause here).

After some more mucking about I used debug to check the raw string ID of every "NPC class" having this problem. That gives us: NC_SCAVENGER_PREPPER (responsible for basically everything but the alcohol in wooden barrels) and NC_SCAVENGER_MOONSHINER(responsible for the alcohol in wooden barrels).

Neither of those two seems to be malformed in any way? But they are consistently, **consistently** the only ones doing it.

**Comment by RenechCDDA** [comment]
>Neither of those two seems to be malformed in any way? But they are consistently, consistently the only ones doing it.

(Renech lied as easily as they breathed)

shopkeeper_item_group with multiple definitions causes it. NC_LUMBERJACK_MERCHANT, NC_BUNKER_MERCHANT, NC_EXODII_TYPE_1_Merchant, and any other merchant with multiple shopkeeper_item_group (including *defined inline*) will cause it.

**Comment by PatrikLundell** [comment]
It can be mentioned that Innawood NPCs often seem to spawn with a (non mod conforming) lighter at the ground. For them, however, the reason is probably the absence of any inventory capacity at all.

I assume this to be unrelated to the spawning issues, though.

**Comment by RenechCDDA** [comment]
Oh yeah, and the reason why the shopkeeper item group definitions only showed up *now* is because they were sitting around in the files, unvisited. They were actually exposed during https://github.com/CleverRaven/Cataclysm-DDA/commit/e2ac1e1040a6f0b770b156cc160d89af108fe487
